subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java

```diff
@@ -109,10 +111,31 @@
     /**
      * Returns a new {@link Provider} from the value of this provider transformed using the given function.
      *
-     * <p>The new provider will be live, so that each time it is queried, it queries this provider and applies the transformation to the result.
-     * Whenever this provider has no value, the new provider will also have no value and the transformation will not be called.
-     *
-     * <p>Any task details associated with this provider are ignored. The new provider will use whatever task details are associated with the return value of the function.</p>
+     * <p>While very similar in functionality to the regular {@link #map(Transformer) map} operation, this method offers a convenient way of chaining
+     * together multiple {@link Provider Providers}, like so:</p>
+     *
+     * <pre>{@code
+     * val configProvider: Provider<Config> = envProvider.flatMap {env ->
```

Comment:
I'm not sure this is the clearest example, because of the nesting.  

I was talking to Sterling yesterday about this method, and he provided the example of a Producer task and a Consumer task, which have both been created using `register` and are typed as `TaskProvider`, where you could use `flatMap` to wire them together like:

`consumerTask.inputFile = producerTask.flatMap { it.outputFile.get() }`

`inputFile` and `outputFile` here would be `RegularFileProperty`s.

This is kind of specific to `TaskProvider`s, but might be a good additional example?

Reply:
Yeah, will try to integrate it. What's sure is that `flatMap` was a really bad choice of name for this method....

Reply:
Gave it a try, pls. repeat your review.

Reply:
I think this is a clearer example.  But I am a little biased :-)